core/collection-factory/mapdb/src/main/java/org/eclipse/rdf4j/collection/factory/mapdb/MapDbCollectionFactory.java

```diff
@@ -0,0 +1,346 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+package org.eclipse.rdf4j.collection.factory.mapdb;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.function.Function;
+
+import org.eclipse.rdf4j.collection.factory.api.BindingSetKey;
+import org.eclipse.rdf4j.collection.factory.api.CollectionFactory;
+import org.eclipse.rdf4j.collection.factory.impl.DefaultCollectionFactory;
+import org.eclipse.rdf4j.common.exception.RDF4JException;
+import org.eclipse.rdf4j.model.Value;
+import org.eclipse.rdf4j.query.BindingSet;
+import org.mapdb.DB;
+import org.mapdb.DBMaker;
+
+public class MapDbCollectionFactory implements CollectionFactory {
+	// The size 16 seems like a nice starting value but others could well
+	// be better.
+	private static final int SWITCH_TO_DISK_BASED_SET_AT_SIZE = 16;
+	protected volatile DB db;
+	protected volatile long colectionId = 0;
+	protected final long iterationCacheSyncThreshold;
+	private final CollectionFactory delegate;
+
+	private static final class RDF4jMapDBException extends RDF4JException {
```

Comment:
Suggestion: instead of this, perhaps introduce a `CollectionFactoryException` that inherits from `RDF4JException`?

Reply:
I am ok with that, the reason I added the RDF4JMapDBException is to make it stand out in the stackdumps as MapDB related. But I can easily add a level in between.